### ISSUE TITLE
Add overlay keyboard with Alt, Ctrl, Tab, arrow keys (and a few more keys) #97

### ISSUE THREAD

**Issue by AliShahrivarian** [termux/termux-app#12]
Hi

It will be good if there be any UI for changing cursor in Termux. It's very annoying to press vol button for simulating `ctrl` key for changing direction (if you have worked with nano you know what I'm saying).

Or it can be simple direction arrow in app or Termux be able to change cursor position on touch.

thank you.

**Comment by fornwall** [comment]
Thanks for the suggestion! There will probably be some kind of optional overlay on the keyboard with additional keys (arrow keys, ctrl/alt/etc, perhaps configurable) like there is in ConnectBot and JuiceSSH in the future! Leaving this issue open to track that.

**Comment by AliShahrivarian** [comment]
Thank you @fornwall

**Comment by fornwall** [comment]
There will probably be something like the JuiceSSH overlay keyboard added soon!

![juicessh-keyboard](https://cloud.githubusercontent.com/assets/277251/11932277/89e862d4-a7f4-11e5-9889-881f01ce3468.png)

**Comment by franciscod** [comment]
please make this optional, I really like the volumekey solution

**Comment by fornwall** [comment]
@franciscod Yes - the plan is to make the overlay keyboard optional - I like the volume key solution too on my phone :).

**Comment by jmunsch** [comment]
+1

**Comment by abjugard** [comment]
Any news on this?

**Comment by renekliment** [comment]
Have you tried using [Hacker's Keyboard](https://f-droid.org/repository/browse/?fdfilter=hacker&fdid=org.pocketworkstation.pckeyboard)?

**Comment by voltagex** [comment]
@fornwall any updates on this?

**Comment by abjugard** [comment]
Problem with using Hacker's Keyboard is having to change keyboard all the time.

**Comment by Neo-Oli** [comment]
Protip: For the time being you can enable arrow keys in Swiftkey, which makes it much more usable with Termux

**Comment by epitron** [comment]
A simple tweak that would help immensely would be to enable the "chrome address bar" keyboard hint, which puts a "/" in place of the "," key.

I switch keyboard modes to hit / quite often in termux.

**Comment by epitron** [comment]
Here's the keyboard hint:
http://developer.android.com/reference/android/widget/TextView.html#attr_android:inputType

textUri:
Text that will be used as a URI. Corresponds to TYPE_CLASS_TEXT | TYPE_TEXT_VARIATION_URI.

**Comment by Neo-Oli** [comment]
This was added in [version 0.35](https://termux.com/changelog.html#termux-v035-2016-08-01) recently. It can be enabled with VolumeUp+Q or long pressing the `keyboard` button in the left drawer menu.

**Comment by toXel** [comment]
The current implementation of the ExtraKeys seems very flexible right? That's what I understand from reading the code. It should be possible to make this completely customizable in the termux.properties file because the buttons are just a String array [here](https://github.com/termux/termux-app/blob/master/app/src/main/java/com/termux/app/ExtraKeysView.java#L107-L109). Of course the button events have to be catched properly but that shouldn't be a big problem.

**Comment by fornwall** [comment]
@toXel Yes - the plan is to make it configurable shortly!

**Comment by AliShahrivarian** [comment]
Sorry for delay :)
+1
It is now more interesting than before.

**Comment by tombh** [comment]
Couldn't find that tip anywhere else on long-pressing the "Keyboard" button in the drawer to toggle the special keys, so I'm commenting here in the hope it gets better documented or I make it more googleable with a wall of relevant text. It was really useful for me because I'm using a hardware keyboard and the Null Keyboard app to stop the system soft keyboard displaying. Even though the Termux special keys dissapear after a second or so when ALT+tabbing back into the app it caused 2 screen resize events which was really annoying in fullscreen terminal apps like tmux - sometimes even, the second resize event wouldn't be honoured, leaving the screen in some weird half resized state :( Anyway, now I can ALT+tab all I want and termux will always looks exactly how I left it :)

**Comment by marksev1** [comment]
I'm definately looking forward to more keys in the overlay, especially the arrow keys I miss.

**Comment by sindastra** [comment]
The discussion on it seems to have started again. I too am in favour to add arrow keys.

There's also a very recent discussion going on here:

https://plus.google.com/114363412663838875988/posts/UkcuMFArGRr

**Comment by mengyyy** [comment]
open SSH  in termux  then connect it with juicessh  
then you will not need termux's keyboard any more.

**Comment by smith153** [comment]
Yea....arrow keys...need them
 Not sure how to use without: (

**Comment by rqcy** [comment]
Please consider the possibility of associating multiple keys with a single button when implementing configurable buttons.  For example, I tend to use Ctrl-c very often and would like to configure a single button for that. I think many people would benefit from that too. Thanks.

**Comment by epitron** [comment]
You can use the Volume-Down button as Ctrl; I never found Ctrl-C annoying to press with that combo.

**Comment by rqcy** [comment]
> You can use the Volume-Down button as Ctrl; I never found Ctrl-C annoying to press with that combo.

I'm aware of the volume-down as ctrl, but I don't find comfortable to
press volume buttons in this context.  You are right, it's not really
annoying, just less comfortable.

**Comment by thinklog** [comment]
mengyyy: "... the connect it with juicessh" 
How? 
I entered ssh at the prompt and got instructions to install openssh.
did that. now when I say ssh it runs openssh. 
Now what? Thanks.

**Comment by Quasic** [comment]
I haven't used it, myself, but I think they were talking about connecting to that ssh server from this app:
https://play.google.com/store/apps/details?id=com.sonelli.juicessh

**Comment by zx2c4** [comment]
The overlay keyboard on JuiceSSH is so clearly superior for many users. Rather than hacks to get the Termux environment running inside JuiceSSH, I'm sure a lot of developers would really appreciate this being added to Termux -- configurable, of course, so as not to alienate users who like the existing kludge.

Is this something the Termux developers are seriously considering? Or does this Issue mainly function as a place for folks like me and others to keep asking?

**Comment by epitron** [comment]
There's a less convoluted way of getting a keyboard that works like juice's:
https://f-droid.org/app/org.pocketworkstation.pckeyboard

It's not identical to juice, but it's got the same basic things, and it's easy to switch to in termux. Might be a good stop-gap.

**Comment by bvgastel** [comment]
Or maybe it is possible to have something like two finger scroll gesture that controls the cursor (both horizontally and vertically). So it won't cost screen real estate but is quickly accessible.

**Comment by knope** [comment]
`shift` would also be nice

**Comment by sdrausty** [comment]
Which file has the code to Ctrl+Q invocation?  I would like to take a closer look at the source code for this issue.  Which files have the code to this issue we are discussing?

**Comment by zx2c4** [comment]
Can this just copy the JuiceSSH overlay 100%? That one works really super well.

**Comment by pro-src** [comment]
One more vote for two-finger gestures to control arrow key functions.

**Comment by sdrausty** [comment]
Links to source code in question?  Please once.  Second request.

**Comment by sobanieca** [comment]
+1 for customizable keys (I need it for Esc, Ctrl, {, }) and gestures handling (I'm using it for Tab key, arrows and PgUp, PgDn, Home, End). Actually, I have switched to Termius app for now as it has this features and allows me to use hardware keyboard (Samsung Galaxy Keyboard Cover). With some help of gestures it works great! But I'm missing whole Termux environment (access to local storage etc.), so it would be great to switch back to Termux ;)

**Comment by Quasic** [comment]
@sdrausty app/src/main/java/com/termux/app/TermuxKeyListener.java has the Ctrl+Q handling, https://github.com/termux/termux-app/commit/d72fd579eea1c7d0bdcab0625d214b0a58d6d1c5 and commits leading up to it built the extra keys view.

**Comment by sdrausty** [comment]
Cool. Thanks.

**Comment by BorhnN** [comment]
+1 for configurable keyboard overlay. 

What's the current update on arrow keys?

Hacker's Keyboard is too congested for my choice. I'm using GBoard with PC layout.

**Comment by mastahcez** [comment]
I am using hackers keyboard but still have to use the overlay because the esc btn is hidden under fn.. overlay customization would be great, something like an rc file for it maybe?

**Comment by agguser** [comment]
Please add arrow keys to Extra Keys Row, or make the row configurable.

**Comment by sdrausty** [comment]
@agguser 
> add arrow keys to Extra Keys Row

Using a touch keyboard at https://termux.com/touch-keyboard has 
 •  Volume Up+Q → Show extra keys view

> make the row configurable

This part is being developed.  See https://raw.githubusercontent.com/termux/termux-app/master/app/src/main/java/com/termux/app/ExtraKeysView.java and https://github.com/termux/termux-app/commit/d72fd579eea1c7d0bdcab0625d214b0a58d6d1c5

**Comment by audiokit-nl** [comment]
Hello Sdrausty, this works super. The extra keys stay on top of the keyboard. Work with emacs a lot and when no external keyboard is attached, the volume up+Q method works great. In terminal mode, the up and down key are used too to browse the former commands, can they be added too? there is some room left :-) Thanks for this great addition! 
Benny

**Comment by neo-expert** [comment]
On my old smartphone the volume keys are broken because I use termux very intensive.
I use this keyboard APP wich has arrow keys:
https://play.google.com/store/apps/details?id=nextapp.inputmethod.latin
But it is not very accurate...

**Comment by sdrausty** [comment]
Hi Benny.  Surprising this pull request, ["add some func keys"](https://github.com/termux/termux-app/pull/605) isn't referenced here.  @neo-expert is the source code available for the app you mention.
# [codeboard](https://play.google.com/store/apps/details?id=com.gazlaws.codeboard)

## Coding on your mobile device just got easier. No more switching between numbers, symbols and alphabets.

### Features:
- Cut, copy, paste, select all
- Tab
- All the braces {[<(
- ; / \ ! | &$ etc
- No ads or IAP, made for students!
# 
*From https://github.com/gazlaws-dev/codeboard

**Comment by jdgregson** [comment]
+1 on this. I can see that the overlay has been released, but it is lacking the arrow keys. Gboard has cursor tools built in now which give you arrow keys, but Termux loads a slimmed down version of Gboard without the arrow key options. I have Hacker's Keyboard installed, but it is really clunky to type on compared to Gboard, and I'd rather not switch to it just to get arrow keys. This really should be a high priority.

**Comment by dwrz** [comment]
Is there any chance we could get horizontal scrolling via touch? It's great to be able to quickly move up and down withh touch, but moving left and right to edit text is always frustrating. As far as I can tell swiping left and right does not do anything at the moment.

**Comment by cy18** [comment]
The current overlay keyboard is not perferct, but quite good. The problem is that it's very hard to find it without Googling. I think it should be turned on by default, or at least should be easier to be found.

**Comment by neo-expert** [comment]
Why is that that difficult to use arrow keys in termux? I am still using an special keyboard app that offers the arrow keys. Unfortunately Gboard does not have these.

**Comment by BorhnN** [comment]
I'd suggest that special keys will have two rows. With the current keys
showing all time with the "on" settings. And a button pointing upward to
show the second row with arrow keys and other necessary keys.

**Comment by BorhnN** [comment]
From @agguser comment I thought I already have this installed on my phone. But then noticed there haven't been a release since January. 

Is there something wrong?

**Comment by diooktput** [comment]
Update to v. 0.62 and you'll be satisfied.

**Comment by Grimler91** [comment]
Closing this after https://github.com/termux/termux-app/pull/605.

**Comment by robertvandeneynde** [comment]
"|" is on the "–" one must long press then go up...
This should really be customizable (not everybody wants two lines, could choose the positions, etc) or at least have a small tutorial somewhere when the app is installed.

**Comment by BorhnN** [comment]
Updated to 0.62 but on my phone this overlay is one row with double the height, I've installed it on another phone and it works fine.

|   | My phone     | Another phone   |
| -------------: |:-------------:| :-----|
| Model | Redmi Note 2 | Redmi 3A |
| Android Version | 5.0 | 6.0 |
| ROM  | MIUI | MIUI |
| Installation | Updated (Google Play) | Installed (Google Play) |
| Scr | [Screenshot Redmi Note 2](https://i.imgur.com/3ND7xeI_d.jpg) | [Screenshot Redmi 3A](https://i.imgur.com/PyYPd71.png) |
| Res | 1080p | 720p |

**Comment by billop** [comment]
How can I disable this? I don't have a navbar and I find it impossible to disable this feature that takes up screen real estate.

**Comment by madacol** [comment]
Here's how to add more special keys
https://wiki.termux.com/wiki/Touch_Keyboard

 ## TL;DR
Add a new file `~/.termux/termux.properties` with a content like this:
```
extra-keys = [ \
 ['ESC','|','/','HOME','UP','END','PGUP','DEL'], \
 ['TAB','CTRL','ALT','LEFT','DOWN','RIGHT','PGDN','BKSP'] \
]
```
![image](https://user-images.githubusercontent.com/6379896/79270461-d6dea900-7e63-11ea-83be-851e4e075493.png)
